### PR TITLE
Mark volume as detached when node does not exist for vsphere

### LINKED ISSUE #50266: Mark volume as detached if instance is deleted from cloud provider.
Cloud provider volume has a function DiskIsAttached to check whether the volume is still attached to the instance. If the instance is no longer exist (deleted by cloud provider), we can safely assume that the volume is no longer attached to the instance and return false without error for DiskIsAttached(). Currently GCE PD and AWS has the correct behavior, but all other volume plugins do not.

### PR DESCRIPTION
**Pull request by FengyunPan**
If node does not exist, node's volumes will be detached
automatically and become available. So mark them detached and
return false without error.
Fix #50266

**Special notes for your reviewer**:
/assign @jingxu97

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/cloudprovider/providers/vsphere/vsphere.go
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: During host failover when node VM goes to disconnected state or is not accessible for any reason we are returning a Map of volumepath and its attachment status set to false. This will be misinterpreted as disks are already detached from the node and Kubernetes mark volumes as detached after orphaned pod is cleaned up. 

This can causes volumes to remain attached to previous node, and new pod creation on another node will remains in the “containercreating” state. Since both the node vms are powered on and hot volumes can not be attached to other node.

We need to test this PR change for vSphere HA failover.

@BaluDontu @luomiao @tusharnt @rohitjogvmw

**Comment by luomiao** [comment]
@FengyunPan Thanks for coming up this PR regarding to the detached volumes.
However, I think this will revert a previous PR to fix another HA issue: https://github.com/kubernetes/kubernetes/pull/45569 
As @divyenpatel has mentioned, directly setting the status of volumes to detached will cause a race condition when the previous dead node is restarted. More explanation can be found in that PR.

Would you please give more information about the reason for this PR?
Do you observe a volume never being detached when a node is removed?
We may need to find a solution to solve both situations.

**Comment by jingxu97** [comment]
@luomiao Could you confirm that when a node is deleted, will the volume be detached automatically? For all other cloud providers such as GCE PD, AWS, that is the behavior. So the function DiskIsAttached() in cloud provider should return false (not attached) without error.

**Comment by luomiao** [comment]
@jingxu97 I feel there are two different issues are mixed here :) but the short answer is no, volumes are not detached automatically.

What @FengyunPan 's PR has changed is not when the node is deleted. Instead, it's when the node is **powered off but still exists** in vCenter. In this case, volumes are still attached to the VM. Thus we cannot simply return false (means disk is not attached). The variable and function names are misleading here. Instead of nodeExist, it should be nodeActive.

Regarding to your question, it also depends on how you define "deleting".
1. If "deleting" means "removing the node from kubernetes cluster":
Then the node is till alive in vCenter. So the volume is still attached to the VM.
2. If "deleting" means "removing the node from vCenter":
Removing node from vCenter involves two steps: power off the node and then delete the node.
When the node is powered-off but not deleted yet, the volumes are attached to the VM.

**Comment by jingxu97** [comment]
@luomiao By "deleted" I mean deleted from vCenter, similar to the case of node instance is deleted from cloud provider.

The code nodeExist, err := vs.NodeExists(nodeName) what does it mean? I thought it means the node is deleted from vcenter already?

**Comment by luomiao** [comment]
@jingxu97 
The current NodeExists is misleading.
It could return nodeExist=false and err=nil, when the node is powered-off but not deleted.
Now I think it's not a problem with this PR but a problem with the NodeExists function.
Please allow us some time to do a few testings to confirm :)
The change to NodeExists function might be required for this PR.

**Comment by luomiao** [comment]
@jingxu97 
Just did some double check. Sorry to take back my words.
NodeExists with a return value nodeExist=false means the node is not in an active status (powered off, disconnected, not found...).
Nodes can be in a non-active state (powered off, disconnected) while volumes are still attached.
In the case of vsphere cloud provider, deleting the node from cloud provider can make the nodes into any of the non-active status (powered off, disconnected, not found...).
But this PR could mark volumes as detached even the volumes might still be attached. Thus it's not correct.

**Comment by jingxu97** [comment]
Is there a way to identify whether the node is deleted from vsphere cloud provider. If no, could we add one?

**Comment by FengyunPan** [comment]
>The variable and function names are misleading here. Instead of nodeExist, it should be nodeActive.
NodeExists with a return value nodeExist=false means the node is not in an active status (powered off, disconnected, not found...).
I see, @luomiao thank for your comments.

>Is there a way to identify whether the node is deleted from vsphere cloud provider. If no, could we add one?
I think it is necessary for attachDetachManager, I'd like to check it.

**Comment by luomiao** [comment]
@FengyunPan @jingxu97 
Sorry for the delayed reply.
The current PR looks good to me.

We will submit another PR to fix NodeExists (NodeActive) function to return InstanceNotFound error when a node is not existing anymore.

/lgtm

**Comment by jingxu97** [comment]
@luomiao should we wait until your fix is merged to merge this PR? Thanks!

**Comment by divyenpatel** [comment]
@jingxu97 @luomiao We should wait for the fix, else vSphere HA feature will be broken.

**Comment by divyenpatel** [comment]
also We have refactored vSphere Cloud Provider at lot. We can address this PR in the new code base once - https://github.com/kubernetes/kubernetes/pull/49164 is merged.

**Comment by luomiao** [comment]
@divyenpatel 
I don't think this PR will break vSphere HA feature though.
This PR has been updated to only return no error when NodeActive function returns err = cloudprovider.InstanceNotFound.
The current NodeActive function doesn't return InstanceNotFound error no matter what state of the node is. Thus it won't change the current code workflow, until we modify the NodeActive function.

Although I agree it's better to hold it after the refactor PR is in :)

**Comment by FengyunPan** [comment]
@luomiao @divyenpatel @jingxu97 I have add codes to check if node exists and rebase the PR ( #49164), Please take a look.

File: pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go]: Not related to the change made in the PR, but just came across this unwanted code block.
I think this is not required.

```
if vmMoList[0].Summary.Config.Template == false {
 		glog.Warningf("VM is not in %s state", ActivePowerState)
 	} else {
 		glog.Warningf("VM is a template")
}
```

File: pkg/cloudprovider/providers/vsphere/vsphere.go
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Returned error should be different. When the Node VM is powered off, we can still determine whether disks are attached to powered off VM or not. Current error message does not look good in this case.

File: pkg/cloudprovider/providers/vsphere/vsphere.go
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Same as commented for DisksAreAttached. When node VM is powered off, we can still check if disks are attached to the Node VM, and return actual state of the disk.

**Review by divyenpatel — CHANGES_REQUESTED** [review]
Provided my review comments.

File: pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go
Comment (FengyunPan) [inline pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go]: I think this change of Exists() is helpful for reader, is right?

File: pkg/cloudprovider/providers/vsphere/vsphere.go
Comment (FengyunPan) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Yes, I agree, there is no need to determine wherher node is powered off in DisksAreAttached() and DisksisAttached().

File: pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go]: Yes `IsActive` is helpful for reader.

**Review by divyenpatel — APPROVED** [review]
Changes looks good to me.

File: pkg/cloudprovider/providers/vsphere/vsphere.go
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: The current log statement is very confusing as when the node is not found, disks will not detached automatically. 
Can you rephrase this info log to
glog.Infof("Node %q does not exist, vsphere CP will assume all disks %v are not attached to it.", nodeName, volPaths).
Same at other places as well.

Comment (FengyunPan) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Ok, thank for your comment.

File: pkg/cloudprovider/providers/vsphere/vsphere.go
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I have one more question. If node is powered off and the pod is placed on to other available nodes, will the volumes attached to powered off node detach automatically?

What logic in here, would cause it to happen? 

Its because i have one PR which does this https://github.com/kubernetes/kubernetes/pull/40118. Throwing an error when node is powered off caused the volume to be detached automatically.

Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Doesn't it matter if the node is powered off or not when the DisksAreAttached() call is made.

Comment (FengyunPan) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: @bacongobbler Have you see the comment from @divyenpatel:"Same as commented for DisksAreAttached. When node VM is powered off, we can still check if disks are attached to the Node VM, and return actual state of the disk."
It seem it doesn't matter if the node is powered off or not when the DisksAreAttached() call is made.
On the other hand, the "vm.Exists(ctx)" just check if vm is power on, can't check if vm is exist.

**Comment by FengyunPan** [comment]
@dims @jingxu97 @kerneltime @luomiao @BaluDontu - Any thoughts on it?

**Comment by BaluDontu** [comment]
The change looks good to me. /lgtm from my side.

**Comment by luomiao** [comment]
sorry for the late reply.
I thought this PR is already merged...
/lgtm
/approve

**Pull request merged**
